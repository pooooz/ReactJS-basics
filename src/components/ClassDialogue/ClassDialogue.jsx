import React from 'react';
import styles from './ClassDialogue.module.scss';
import { ClassMessageList } from './components/ClassMessageList/ClassMessageList';
import { Input } from './components/Input/Input';

export class ClassDialogue extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      messages: [{ text: 'Initialization...', author: 'Admin' }],
      inputValue: '',
      botMessageCounter: 0,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.timeout = null;
  }

  handleChange(event) {
    this.setState({ inputValue: event.target.value });
  }

  handleSubmit(event) {
    if (this.state.inputValue !== '') {
      this.state.messages.push({ text: this.state.inputValue, author: 'You' });
      this.setState({
        messages: this.state.messages,
        inputValue: '',
        botMessageCounter: this.state.botMessageCounter + 1,
      });
    }
    event.preventDefault();
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.botMessageCounter !== prevState.botMessageCounter) {
      clearTimeout(this.timeout);
      this.timeout = setTimeout(() => {
        this.state.messages.push({ text: '[BOT] Message', author: 'BOT' });
        this.setState({ messages: this.state.messages });
      }, 1500);
    }
  }

  render() {
    return (
      <section className={styles.dialogue}>
        <ClassMessageList messages={this.state.messages} />
        <form className={styles.form} onSubmit={this.handleSubmit}>
          <Input value={this.state.inputValue} change={this.handleChange} />
          <button className={styles.form__button} type="submit">
            Send
          </button>
        </form>
      </section>
    );
  }
}
