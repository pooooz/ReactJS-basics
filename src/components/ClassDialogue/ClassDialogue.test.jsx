import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/user-event';
import { ClassDialogue } from './ClassDialogue';

describe('ClassDialogue', () => {
  it('Render component', () => {
    render(<ClassDialogue />);
  });

  it('Render with snapshot', () => {
    const { asFragment } = render(<ClassDialogue />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Input test', () => {
    render(<ClassDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    expect(screen.getByRole('textbox')).toContainHTML('Test');
  });

  it('Forming a message', () => {
    render(<ClassDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByText(/<You>/)).toBeInTheDocument();
  });

  it('Forming a empty message', () => {
    const { asFragment } = render(<ClassDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('Bot answer', async () => {
    render(<ClassDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(await screen.findByText(/\[BOT\] Message/)).toBeInTheDocument();
  });
});
