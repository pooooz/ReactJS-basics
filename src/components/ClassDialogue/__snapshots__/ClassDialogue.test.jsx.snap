// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassDialogue Forming a empty message 1`] = `
<DocumentFragment>
  <section>
    <div>
      <div>
        <p
          data-text="Initialization..."
        >
          Initialization...
        </p>
        <p
          data-text="Author: <Admin>"
        >
          Author: &lt;Admin&gt;
        </p>
      </div>
    </div>
    <form>
      <div>
        <input
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Send
      </button>
    </form>
  </section>
</DocumentFragment>
`;

exports[`ClassDialogue Render with snapshot 1`] = `
<DocumentFragment>
  <section>
    <div>
      <div>
        <p
          data-text="Initialization..."
        >
          Initialization...
        </p>
        <p
          data-text="Author: <Admin>"
        >
          Author: &lt;Admin&gt;
        </p>
      </div>
    </div>
    <form>
      <div>
        <input
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Send
      </button>
    </form>
  </section>
</DocumentFragment>
`;
